File: src/main/java/com/stuypulse/robot/subsystems/Feeder/FeederImpl.java
package com.stuypulse.robot.subsystems.Feeder;

import com.ctre.phoenix6.controls.Follower;
import com.ctre.phoenix6.controls.VelocityVoltage;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.signals.MotorAlignmentValue;
import com.stuypulse.robot.constants.Motors;
import com.stuypulse.robot.constants.Ports;

import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class FeederImpl extends Feeder {
    private final TalonFX feederMotor1;
    private final TalonFX feederMotor2;

    public FeederImpl() {
        feederMotor1 = new TalonFX(Ports.Feeder.FEEDER_MOTOR_1);
        feederMotor2 = new TalonFX(Ports.Feeder.FEEDER_MOTOR_2);

        Motors.Feeder.MOTOR_CONFIG_1.configure(feederMotor1);
        Motors.Feeder.MOTOR_CONFIG_2.configure(feederMotor2);

        feederMotor2.setControl(new Follower(Ports.Feeder.FEEDER_MOTOR_1, MotorAlignmentValue.Opposed)); //TODO: figure out motor alignment
    }

    public void setMotors() {
        feederMotor1.setControl(new VelocityVoltage(getState().getTargetRPM()));
    }

    @Override
    public void periodic() {
        super.periodic();
        setMotors();

        SmartDashboard.putNumber("Feeder/Target RPM", getState().getTargetRPM());
        SmartDashboard.putNumber("Feeder/Current RPM", feederMotor1.getVelocity().getValueAsDouble());
    }
}

File: src/main/cpp/subsystems/feeder/FeederImpl.cpp
#include "subsystems/feeder/FeederImpl.hpp"
#include "constants/Motors.hpp"

FeederImpl::FeederImpl() {
    Motors::Feeder::LEADER_CONFIG.configure(feederMotor);
}

void FeederImpl::Periodic() {
    // Control
    feederMotor.SetControl(ctre::phoenix6::controls::DutyCycleOut(getState().getTargetDutyCycle()));

    // Logging
    tkit::Logger& logger = tkit::Logger::GetInstance();
    logger.RecordOutput("Feeder/Velocity_RPS", feederMotor.GetVelocity().GetValueAsDouble());
    logger.RecordOutput("Feeder/Stator Current_Amps", feederMotor.GetStatorCurrent().GetValueAsDouble());
    logger.RecordOutput("Feeder/Supply Current_Amps", feederMotor.GetSupplyCurrent().GetValueAsDouble());

    Feeder::Periodic();
}